### PR TITLE
lab 03 updated, issue fixed

### LINKED ISSUE #447: Lab 03 Letter Frequency
/* Issue to be closed when pull request approved and merged */

Create an executable go program in directory 03_letters/marchitect
Write a function that returns a mapping of each letter to its frequency
Bonus points: comprehensive tests

### PR DESCRIPTION
**Pull request by Marchitect**
Fixes #447

Review of colleague's PR #

Changes proposed in this PR:
- added function letters() and sortLetters(), test cases, issue raised by bot hopefully fixed

### CODE REVIEW

**Review by nicko-lee — APPROVED** [review]
Interesting logic of implementing letters func - didn't think of approaching it this way

**Comment by Marchitect** [comment]
> Interesting logic of implementing letters func - didn't think of approaching it this way

I wanted to say I don't even understand my logic now lolz, thank you for the feedback though :)

File: 03_letters/marchitect/main.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	"sort"
+	"strconv"
+	"strings"
+)
+
+var out io.Writer = os.Stdout
+
+func letters(s string) map[rune]int {
+	m := make(map[rune]int)
+	// keep a list of what characters exist in the string s
```
Comment (juliaogris) [inline 03_letters/marchitect/main.go:16]: what for?

File: 03_letters/marchitect/main.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	chars := ""
+	for _, c := range s {
+		if !strings.ContainsAny(chars, string(c)) {
```
Comment (juliaogris) [inline 03_letters/marchitect/main.go:19]: This is inefficient. ContainsAny is `O(n)` and so it `strings.Count(s, string(c))`, updating a map is `O(1)`.
I accept that runtime is irrelevant for the size of strings expected in this program, but it is good practise to be aware. 
```go
for _, c := range s {
    m[c]++
}
```
should do the trick

File: 03_letters/marchitect/main.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+			m[c] = numberOfc
+			chars += string(c)
+		}
+	}
+	return m
+}
+
+func sortLetters(m map[rune]int) []string {
+	result := make([]string, 0)
+	keys := make([]int, 0)
```
Comment (juliaogris) [inline 03_letters/marchitect/main.go:30]: Add the capacity to make as you know it `len(m)`.

File: 03_letters/marchitect/main.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	for k := range m {
+		keys = append(keys, int(k))
+	}
+	sort.Ints(keys)
+	for _, k := range keys {
+		unquoted, _ := strconv.Unquote(strconv.QuoteRune(rune(k)))
+		result = append(result, unquoted+":"+strconv.Itoa(m[rune(k)]))
```
Comment (juliaogris) [inline 03_letters/marchitect/main.go:37]:
```go
s := fmt.Sprintf("%c:%d", rune(k), m[rune(k)])
result = append(result, s)
```

File: 03_letters/marchitect/main_test.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+		{"normal test 2", "$Bfs`BY", []string{"$:1", "B:2", "Y:1", "`:1", "f:1", "s:1"}},
+		{"empty input", "", []string{}},
+		{"Mandarin Characters Hello", "你好", []string{"你:1", "好:1"}},
+		{"normal test 3", "Brain drained", []string{" :1", "B:1", "a:2", "d:2", "e:1", "i:2", "n:2", "r:2"}},
+	}
+	for _, table := range tables {
+		expected := sortLetters(letters(table.input))
+		actual := table.output
+		// reference: https://yourbasic.org/golang/compare-slices/
+		if !reflect.DeepEqual(actual, expected) {
```
Comment (juliaogris) [inline 03_letters/marchitect/main_test.go:27]: Use testify assert.Equal to turn these three lines into a oneliner.

**Review by juliaogris — CHANGES_REQUESTED** [review]
Please add link to your review of colleague's PR and [fix commit messages](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments#commit-messages) with `git rebase -i HASH`

**Pull request closed without merging**